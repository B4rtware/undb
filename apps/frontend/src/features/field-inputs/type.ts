import type { IOptionColor } from '@undb/core'

export type OnColorChange = (color: IOptionColor) => void
