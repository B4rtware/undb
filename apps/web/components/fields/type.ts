import type { IOptionColor } from '@egodb/core'

export type OnColorChange = (color: IOptionColor) => void
