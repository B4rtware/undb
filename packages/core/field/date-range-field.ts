import type { IDateRangeFilterOperator } from '../filter'
import type { IDateRangeFilter } from '../filter/date-range.filter'
import { DateRangeFieldValue } from './date-range-field-value'
import type {
  DateRangeType,
  ICreateDateRangeFieldSchema,
  ICreateDateRangeFieldValue,
  IDateRangeFieldValue,
} from './date-range-field.type'
import { BaseField } from './field.base'
import type { IDateRangeField } from './field.type'
import { FieldId, FieldName, FieldValueConstraints } from './value-objects'

export class DateRangeField extends BaseField<IDateRangeField> {
  get type(): DateRangeType {
    return 'date-range'
  }

  static create(input: ICreateDateRangeFieldSchema): DateRangeField {
    return new DateRangeField({
      id: FieldId.from(input.id),
      name: FieldName.create(input.name),
      valueConstrains: FieldValueConstraints.create({ required: input.required }),
    })
  }

  static unsafeCreate(input: ICreateDateRangeFieldSchema): DateRangeField {
    return new DateRangeField({
      id: FieldId.from(input.id),
      name: FieldName.unsafaCreate(input.name),
      valueConstrains: FieldValueConstraints.unsafeCreate({ required: input.required }),
    })
  }

  createValue(value: ICreateDateRangeFieldValue): DateRangeFieldValue {
    return new DateRangeFieldValue(value)
  }

  createFilter(operator: IDateRangeFilterOperator, value: IDateRangeFieldValue | null): IDateRangeFilter {
    return { operator, value: value, path: this.name.value, type: 'date-range' }
  }
}
