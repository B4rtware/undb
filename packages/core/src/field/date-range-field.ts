import type { IDateRangeFilter } from '../filter/date-range.filter.js'
import type { IDateRangeFilterOperator } from '../filter/index.js'
import { DateRangeFieldValue } from './date-range-field-value.js'
import type { DateRangeType, ICreateDateRangeFieldSchema, ICreateDateRangeFieldValue } from './date-range-field.type.js'
import { BaseDateField } from './field.base.js'
import type { IDateRangeField } from './field.type.js'
import type { IFieldVisitor } from './field.visitor.js'
import { DateFormat, FieldId, FieldName, FieldValueConstraints } from './value-objects/index.js'

export class DateRangeField extends BaseDateField<IDateRangeField> {
  type: DateRangeType = 'date-range'

  override get primitive() {
    return true
  }

  static create(input: Omit<ICreateDateRangeFieldSchema, 'type'>): DateRangeField {
    const fieldName = FieldName.create(input.name)

    return new DateRangeField({
      id: FieldId.fromNullableString(input.id),
      name: fieldName,
      valueConstrains: FieldValueConstraints.create({ required: input.required }),
      format: input.format ? DateFormat.fromString(input.format) : undefined,
    })
  }

  static unsafeCreate(input: ICreateDateRangeFieldSchema): DateRangeField {
    return new DateRangeField({
      id: FieldId.fromNullableString(input.id),
      name: FieldName.unsafaCreate(input.name),
      valueConstrains: FieldValueConstraints.unsafeCreate({ required: input.required }),
      format: input.format ? DateFormat.fromString(input.format) : undefined,
    })
  }

  createValue(value: ICreateDateRangeFieldValue): DateRangeFieldValue {
    return new DateRangeFieldValue(value)
  }

  createFilter(operator: IDateRangeFilterOperator, value: [string | null, string | null] | null): IDateRangeFilter {
    return {
      operator,
      value: value ?? [null, null],
      path: this.id.value,
      type: 'date-range',
    }
  }

  accept(visitor: IFieldVisitor): void {
    visitor.dateRange(this)
  }
}
