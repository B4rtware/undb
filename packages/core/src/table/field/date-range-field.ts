import { andOptions } from '@undb/domain'
import type { Option } from 'oxide.ts'
import { z } from 'zod'
import type { IDateRangeFilter } from '../filter/date-range.filter.js'
import type { IDateRangeFilterOperator } from '../filter/index.js'
import type { TableCompositeSpecificaiton } from '../specifications/interface.js'
import { DateRangeFieldValue } from './date-range-field-value.js'
import type {
  DateRangeType,
  ICreateDateRangeFieldSchema,
  IDateRangeFieldQueryValue,
  IUpdateDateRangeFieldInput,
} from './date-range-field.type.js'
import { AbstractDateField } from './field.base.js'
import type { IDateRangeField } from './field.type.js'
import type { IFieldVisitor } from './field.visitor.js'
import { DateFormat } from './value-objects/date-format.vo.js'
import { FieldId } from './value-objects/field-id.vo.js'
import { TimeFormat } from './value-objects/time-format.vo.js'

export class DateRangeField extends AbstractDateField<IDateRangeField> {
  duplicate(name: string): DateRangeField {
    return DateRangeField.create({
      ...this.json,
      id: FieldId.createId(),
      name,
    })
  }

  type: DateRangeType = 'date-range'

  override get json() {
    return {
      ...super.json,
      format: this.formatString,
      timeFormat: this.timeFormatString,
    }
  }

  override get primitive() {
    return true
  }

  static create(input: Omit<ICreateDateRangeFieldSchema, 'type'>): DateRangeField {
    return new DateRangeField({
      ...super.createBase(input),
      timeFormat: input.timeFormat ? TimeFormat.from(input.timeFormat) : undefined,
      format: input.format ? DateFormat.fromString(input.format) : undefined,
    })
  }

  static unsafeCreate(input: ICreateDateRangeFieldSchema): DateRangeField {
    return new DateRangeField({
      ...super.unsafeCreateBase(input),
      timeFormat: input.timeFormat ? TimeFormat.from(input.timeFormat) : undefined,
      format: input.format ? DateFormat.fromString(input.format) : undefined,
    })
  }

  public override update(input: IUpdateDateRangeFieldInput): Option<TableCompositeSpecificaiton> {
    return andOptions(this.updateBase(input), this.updateFormat(input.format), this.updateTimeFormat(input.timeFormat))
  }

  createValue(value: IDateRangeFieldQueryValue): DateRangeFieldValue {
    return DateRangeFieldValue.fromQuery(value)
  }

  createFilter(operator: IDateRangeFilterOperator, value: [string | null, string | null] | null): IDateRangeFilter {
    return {
      operator,
      value: value ?? [null, null],
      path: this.id.value,
      type: 'date-range',
    }
  }

  accept(visitor: IFieldVisitor): void {
    visitor.dateRange(this)
  }

  get valueSchema() {
    const dateRange = z.tuple([z.string().nullable(), z.string().nullable()])

    return this.required ? dateRange : dateRange.nullable()
  }
}
