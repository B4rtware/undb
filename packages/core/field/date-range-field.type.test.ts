import { addMilliseconds } from 'date-fns'
import { SafeParseError } from 'zod'
import { dateRangeFieldValue, IDateRangeFieldValue } from './date-range-field.type'

const now = new Date()

test.each<[IDateRangeFieldValue, boolean]>([
  [{ from: now, to: addMilliseconds(now, 1) }, false],
  [{ from: now, to: addMilliseconds(now, -1) }, true],
  [{ from: now, to: now }, true],
  [null, false],
])('dateRangeFieldValue', (data, expectError) => {
  const result = dateRangeFieldValue.safeParse(data)

  if (!expectError) {
    expect(result.success).toBe(true)
  } else {
    expect((result as SafeParseError<boolean>).error).toMatchSnapshot()
  }
})
