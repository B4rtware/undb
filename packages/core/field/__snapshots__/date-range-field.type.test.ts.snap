// Vitest Snapshot v1

exports[`dateRangeFieldValue 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "date range value from should before value to",
    "path": []
  }
]]
`;

exports[`dateRangeFieldValue 2`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "date range value from should before value to",
    "path": []
  }
]]
`;
